### ISSUE TITLE
  Neovim refuses to go to background with Ctrl-Z

### ISSUE THREAD

**Issue by roblabla** [neovim/neovim#3100]
Sending neovim to background with ctrl-z does not work on my setup. I also tried running neovim with `nvim -u NORC` to disable my plugin, config and whatnot, sending to background is still broken. It works correctly under normal vim.

OS : ArchLinux x64
Terminal : PuTTy
Shell : zsh, bash (tried both)

```
nvim -v
NVIM 0.0.0-alpha+201507232235 (compiled Jul 25 2015 12:14:21)
Commit: 7a6bf3f418c5ad94ac2ac71f21275a87d08e87b9
Build type: Release
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong --param=ssp-buffer-size=4 -Wconversion -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/tmp2/neovim-git/src/neovim-git/build/config -I/tmp2/neovim-git/src/neovim-git/src -I/usr/include -I/usr/include -I/usr/include/luajit-2.0 -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/tmp2/neovim-git/src/neovim-git/build/src/nvim/auto -I/tmp2/neovim-git/src/neovim-git/build/include
Compiled by roblabla@roblab.la

Optional features included (+) or not (-): +acl   +iconv    +jemalloc
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/usr/share/nvim"
```

**Comment by fmoralesc** [comment]
This might be relevant: http://the.earth.li/~sgtatham/putty/0.58/htmldoc/Chapter4.html#config-telnetkey

**Comment by roblabla** [comment]
After some investigations, `TERM=xterm-256color nvim` lets me Ctrl-Z properly. I made my putty report the `putty-256color` termcaps by default, which seem sto make Ctrl-Z fail to work.

**Comment by Grimy** [comment]
I can confirm that the following will not background nvim:

* `env TERM=putty-256color nvim -u NONE`
* Hit <kbd>Ctrl-Z</kbd>

However, I believe it is the correct behavior, considering that Ctrl-Z isn’t enabled by default in Putty. Sure, setting `TERM` incorrectly breaks things, but that’s not new.

**Comment by roblabla** [comment]
With putty-256color set, I haven't met a single app that refused to go to background, except for neovim. It's ok since hacking the TERM variable fixes it, but that isn't the expected behavior, regardless of whether it is correct or not (I have long abandoned the idea of correctness when it came to terminal emulation...)

**Comment by crondog** [comment]
This also happens when TERM=linux

Running nvim in tty2 does this and so does "TERM=linux nvim" from inside xterm and st

**Comment by crondog** [comment]
In fact this happens with any terminfo which contains kspd

```
infocmp linux | grep kspd
kmous=\E[M, knp=\E[6~, kpp=\E[5~, kspd=^Z, nel=^M^J, oc=\E]R,

infocmp putty-256color | grep kspd
kmous=\E[M, knp=\E[6~, kpp=\E[5~, kri=\EOA, kspd=^Z,
```

However these work

```
infocmp xterm | grep kspd
# no results found

infocmp st | grep kspd
# no results found
```

Not sure of what is exactly causing the issue though

**Comment by wsdjeg** [comment]
now I can not use Ctrl+z send neovim to background,It works in well in vim ,my os is ubuntu ,and I am using gnome-terminal

**Comment by justinmk** [comment]
@wsdjeg Please open a new issue with precise details for new issues. On existing issues, please only leave new, relevant information.

**Comment by wsdjeg** [comment]
@justinmk ok

**Comment by wsdjeg** [comment]
@justinmk now Ctrl+z works well for me ,but it will save my file before goto background ,sometimes i do not want neovim save my file .

**Comment by justinmk** [comment]
@wsdjeg you keep reporting various symptoms without mentioning any changes in your environment. There is nothing we can do with this information...

**Comment by wsdjeg** [comment]
@justinmk I am using the daily build package,from https://launchpad.net/~neovim-ppa/+archive/ubuntu/unstable,two days ago ,I can the when the package was builded by `nvim --version`,but now the the feature has been moved,also other thing does not changed.
my os is ubuntu 16.04, and my terminal is gnome-terminal version 3.18.2

**Comment by wsdjeg** [comment]
@justinmk sorry ,I have test this feature in vim ,now everything works same as vim ,so I have no issue with this feature,maybe @roblabla has different  environment with me .

**Comment by dontdieych** [comment]
I have same problem in Arch linux console (TERM=linux).

nvim 0.1.1

**Comment by dontdieych** [comment]
No response from c-z after c-v in insert mode and command mode.

**Comment by akavel** [comment]
Also have this issue, in NixOS. 

- in Alt-F1 etc. consoles, with TERM=linux, Ctrl-Z doesn't work. (It works correctly with the same neovim in xterm in Alt-F7 graphical console, where TERM=xterm.)
- when I run `:suspend` or `:stop`, they execute successfully and throw me to shell, where `fg` restores nvim;
- in insert mode, nothing seems to happen on Ctrl-Z either; even if I do Ctrl-V, pressing Ctrl-Z still does nothing (a `^` is still displayed, waiting for some keypress);
- I tried adding a signal handler for SIGTSTP in src/nvim/os/signal.c, printing some text to a file, but it was not called (the file was not created) on Ctrl-Z; the same code inserted on first line of signal_init resulted in the file being created, so it was not a bug in the debugging code;
- I tried `kill -TSTP $PID` from another console; it stopped neovim, but in a weird way: the `[1]+ Stopped   bin/nvim` message got printed at the top of the screen, and vim's screen stayed visible "in background" (instead of vim's screen scrolling up and the "Stopped" message showing below it). Then after `fg`, vim doesn't register keys immediately, only after <CR> (even Ctrl-L gets handled only after a subsequent <CR>).

It looks to me like the keypress doesn't get converted into SIGTSTP somehow? weird. No idea what can do it, such that even a registered SIGTSTP handler doesn't get called! Also, for other programs (e.g. less), Ctrl-Z works fine.

**Comment by ZyX-I** [comment]
@akavel Ctrl-Z was not ever supposed to be converted to SIGTSTP, Vim and, after it, Neovim was always supposed to handle keys in order (which is especially relevant given that `<C-z>` meaning depends on mode in which it is used, and this meaning is not necessary “suspend Neovim”), so `<C-z>` is not supposed to be handled up until previous input was handled, which is much easier to do if you are not using signals.

It would be weird if signal handler *was* called, this is simply unnecessary complication: in this handler you would need to push `<C-z>` into input queue. Keys like `<C-s>`, `<C-z>`, `<C-q>`, `<C-d>` (BTW, this one also does not mean EOF for (Neo)Vim) are easy to disable using `stty`, for `<C-z>` this would be `stty susp ''` (and, of course, everybody is free to use the same method stty is using on its own to avoid spawning a process).

**Comment by akavel** [comment]
The PR I sent above seems to fix it for me, and still work in xterm.

**Comment by coderextreme** [comment]
Environment:

Windows 10
Windows Terminal
Git Bash from "Git For Windows"
Neovim 0.9.5 from GitHub release, not master source.
```
$ env |grep TERM
TERM=xterm-256color
``
CTRL-Z in normal mode does nothing.

Expected behavior:
Drop into the shell, and can resume with shell fg command.

vim from Git 4 Windows works ok
